### PR TITLE
parse .debug_aranges section

### PR DESCRIPTION
**Pull request by dorothychen**
parse aranges table in dwarf and provide simple interface to get a CU offset from an address

### CODE REVIEW

File: elftools/dwarf/aranges.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        self.stream = stream
+        self.size = size
+        self.structs = structs
+
+        # entries is a sorted list of ARangeEntry tuples, 
+        # sorted by beginning address
+        self.entries = self._get_entries()
+        self.entries.sort(key=lambda entry: entry[0])
+        self.keys = [entry[0] for entry in self.entries]
+
```
Comment (eliben) [inline elftools/dwarf/aranges.py]: here and elsewhere: please one blank line between class methods

File: elftools/dwarf/aranges.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+import os
+from collections import namedtuple
+from ..common.utils import struct_parse
+from bisect import bisect_left
+import math
+
+ARangeEntry = namedtuple('ARangeEntry', 'begin_addr length info_offset')
+
+class ARanges(object):
+    """ ARanges table in DWARF
```
Comment (eliben) [inline elftools/dwarf/aranges.py]: Lacks comments: how to use this class? Methods and ctor need usage comments too

File: elftools/dwarf/aranges.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+                addr = struct_parse(addr_size('addr'), self.stream)
+                length = struct_parse(addr_size('length'), self.stream)
+                while addr != 0 and length != 0:
+                    entries.append(
+                        ARangeEntry(begin_addr=addr, 
+                            length=length, 
+                            info_offset=aranges_header["debug_info_offset"]))
+                    addr = struct_parse(addr_size('addr'), self.stream)
+                    length = struct_parse(addr_size('length'), self.stream)
+
```
Comment (eliben) [inline elftools/dwarf/aranges.py]: no blank lines between if/elif bodies

File: elftools/dwarf/aranges.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+            elif segment_size != 0:
+                raise NotImplementedError("Segmentation not supported")
+
+            offset = (offset 
+                + aranges_header.unit_length 
+                + self.structs.initial_length_field_size())
+
+        return entries
+
+    def _get_addr_size(self, addr_header_value):
```
Comment (eliben) [inline elftools/dwarf/aranges.py]: I'd rename this to `_get_addr_size_struct` or something of the sort to make it clear it's a struct that should be parsed; same for the usages

**Comment by eliben** [comment]
Thanks for the contribution!

I left some inline comments. Also, this is missing tests. I'm pretty sure many of the existing test objects/binaries for readelf-comparison tests already have aranges sections so you can enable this section in comparisons and make sure the tests exercise it and everything works.

**Comment by dorothychen** [comment]
Noted! I've made the cosmetic changes and added aranges to the readelf tests. 

The formatting of readelf/greadelf might be slightly different on Mac vs linux (some of the original tests fail because of extra or too few zeros on my machine)

**Comment by dorothychen** [comment]
Also, does it make sense to have a method in ARanges that returns a CU given an address? I am not quite sure if it's more consistent with the library to return an offset or the actual object.

**Comment by eliben** [comment]
> Also, does it make sense to have a method in ARanges that returns a CU given an address? I am not quite sure if it's more consistent with the library to return an offset or the actual object.

Good question. I wonder how do you intend to use it? Are you going to call `DWARFInfo._parse_CU_at_offset` after getting the offset from `ARanges`?

**Comment by dorothychen** [comment]
> Good question. I wonder how do you intend to use it? Are you going to call DWARFInfo._parse_CU_at_offset after getting the offset from ARanges?

Yes, as of now all my use cases involve getting the CU offset and immediately parsing the CU at that offset

**Comment by eliben** [comment]
So I'd leave it as it is for now. In general, it probably makes sense to expose `_parse_CU_at_offset` as a public method of `DWARFInfo`.

You can go ahead and squash all commits in this PR into a single one, and I'll merge it.

**Comment by dorothychen** [comment]
squashed/rebased

**Pull request merged**